### DISCUSSION TITLE
Is there a way to subscribe to form changes that doesn't trigger re-renders?

### DISCUSSION THREAD

**Discussion by alexkatz** [TanStack/form#515] [General]
For example, in react-hook-form, it's possible to do something like:

```
useEffect(() => {
  const sub = watch(formData => { 
    console.log('Side effect'); 
  });

  return sub.unsubscribe;
}, [watch]);
```

I've noticed `FormApi`'s `Subscribe` component and the `useStore` hook which both seem handy, but don't quite cover this use case.

Perhaps `onChange` and `onChangeAsync` handle this, but they seem geared more toward validation and I'm hesitant to put side effects in there.

Are there plans for this kind of functionality? Maybe I'm overlooking something?

Anyway, thanks for such an awesome library!

**Comment by alexkatz** [comment]
Welp, answered my own question after fiddling around for a little:

```
useEffect(() => {
  return form.store.subscribe(() => {
    console.log(form.store.state.values);
  });
}, [form.store]);
```

This allows me to observe changes to form state without triggering any re-renders. 

It's not documented on the TanStack Form website, but it *is* documented in the underlying [TanStack Store docs](https://tanstack.com/store/latest/docs/reference/Store).

**Reply by crutchcorn** [reply]
Any chance you'd want to open a PR to document this for others? :) Our docs need some help!

**Reply by alexkatz** [reply]
Sure! [Here it is.](https://github.com/TanStack/form/pull/517#issue-1986389416)

I figured it's easiest to link directly to the Store docs from the place where the Store instance is mentioned in the Form docs.

Feel free to alter it however you see fit.

**Reply by xmsz-stu** [reply]
hi is any way to subscibe only one field, or how can i know which field is changed?

**Reply by Balastrong** [reply]
> hi is any way to subscibe only one field, or how can i know which field is changed?

Have a look at the [Reactivity](https://tanstack.com/form/latest/docs/framework/react/guides/basic-concepts#reactivity) docs :)

**Reply by xmsz-stu** [reply]
> > hi is any way to subscibe only one field, or how can i know which field is changed?
> 
> Have a look at the [Reactivity](https://tanstack.com/form/latest/docs/framework/react/guides/basic-concepts#reactivity) docs :)

But it will case re-render

**Reply by Balastrong** [reply]
Yes, subscribing to a field is designed for that, to trigger a rerender when the value is updated.

**Reply by xmsz-stu** [reply]
> Yes, subscribing to a field is designed for that, to trigger a rerender when the value is updated.

ok

**Comment by JohanCodinha** [comment]
Is this the way to go to trigger side effect on form store change ?
I want to update url query params when the form values change.

**Reply by nico-martinucci** [reply]
Alternative could be to do it inside a form inputs `onChange`, depending on how many you have to manage.
